plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/state/FollowingViewportStateImpl.kt

```diff
@@ -0,0 +1,238 @@
+    lastLocation = point
+    notifyLatestViewportData()
+  }
+
+  private val indicatorBearingChangedListener = OnIndicatorBearingChangedListener { bearing ->
+    if (options.bearing == FollowingViewportStateBearing.SyncWithLocationPuck) {
+      lastBearing = bearing
+      notifyLatestViewportData()
+    }
+  }
+
+  private fun notifyLatestViewportData() {
+    val viewportData = evaluateViewportData()
+    if (isFollowingStateRunning) {
+      // Use instant update here since the location updates are already interpolated by the location component plugin
+      updateFrame(viewportData, true)
+    }
+    dataSourceUpdateObservers.forEach {
+      if (!it.onNewData(viewportData)) {
+        dataSourceUpdateObservers.remove(it)
+      }
+    }
+  }
+
+  private fun evaluateViewportData(): CameraOptions {
+    return CameraOptions.Builder()
+      .center(lastLocation)
+      .bearing(
+        with(options.bearing) {
+          when (this) {
+            is FollowingViewportStateBearing.Constant -> bearing
+            else -> lastBearing
+          }
+        }
+      )
+      .zoom(options.zoom)
+      .pitch(options.pitch)
+      .padding(options.padding)
+      .build()
+  }
+
+  private fun addIndicatorListenerIfNeeded() {
+    if (!isObservingLocationUpdates) {
+      locationComponent.addOnIndicatorPositionChangedListener(indicatorPositionChangedListener)
+      locationComponent.addOnIndicatorBearingChangedListener(indicatorBearingChangedListener)
+      isObservingLocationUpdates = true
+    }
+  }
+
+  private fun removeIndicatorListenerIfNeeded() {
+    if (isObservingLocationUpdates && dataSourceUpdateObservers.isEmpty() && !isFollowingStateRunning) {
+      locationComponent.removeOnIndicatorPositionChangedListener(indicatorPositionChangedListener)
+      locationComponent.removeOnIndicatorBearingChangedListener(indicatorBearingChangedListener)
+      isObservingLocationUpdates = false
+    }
+  }
+
+  /**
+   * Describes the configuration options of the state.
+   */
+  override var options: FollowingViewportStateOptions = initialOptions
+    set(value) {
+      field = value
+      notifyLatestViewportData()
+    }
+
+  /**
+   * Observer the new camera options produced by the [ViewportState], it can be used to get the
+   * latest state [CameraOptions] for [ViewportTransition].
+   *
+   * @param viewportStateDataObserver observer that observe new viewport data.
+   * @return a handle that cancels current observation.
+   */
+  override fun observeDataSource(viewportStateDataObserver: ViewportStateDataObserver): Cancelable {
+    checkLocationComponentEnablement()
+    addIndicatorListenerIfNeeded()
+    dataSourceUpdateObservers.add(viewportStateDataObserver)
+    return Cancelable {
+      dataSourceUpdateObservers.remove(viewportStateDataObserver)
+      removeIndicatorListenerIfNeeded()
+    }
+  }
+
+  private fun checkLocationComponentEnablement() {
+    if (!locationComponent.enabled) {
+      Logger.w(
```

Comment:
Perhaps throw runtime exception? Seems like a severe violation to me

Reply:
btw - why do we even need this? our viewport plugin requires location + camera as implementations, I guess we should rather have a check that location component is _enabled_.
or alternatively similar logic with similar warning should be added for camera plugin dependency as well

Reply:
I didn't make it a fatal error as user could decide to enable the location component later, then the viewport will sync with the puck. I see it more as a warning that the following viewport data source has stopped producing data.